src/core/net/dns_headers.hpp

```diff
@@ -374,6 +374,243 @@ class Header : public Clearable<Header>
 
 } OT_TOOL_PACKED_END;
 
+/**
+ * This class implement helper methods for encoding/decoding of DNS Domain Names.
+ *
+ */
+class DomainName
+{
+public:
+    enum : uint8_t
+    {
+        kMaxLabelLength = 63,  ///< Max number of characters in a label.
+        kMaxLength      = 255, ///< Max number of characters in a domain name.
+    };
+
+    /**
+     * This static method encodes and appends a single domain name label to a message.
+     *
+     * The @p aLabel is assumed to contain a single domain name label as a C string (null-terminated). Unlike
+     * `AppendMultipleLabels()` which parses the label string and treats it as sequence of multiple (dot-separated)
+     * labels, this method always appends @p aLabel as a single whole label. This allows the label string to even
+     * contain dot '.' character, which, for example, is useful for "Service Instance Names" where <Instance> portion
```

Comment:
nit: I agree it would be desirable to reuse these methods for other related things. Note, however, that Service Instances Names include Domain Names, but Service Instances Names are not Domain Names in themselves. Should we move more general methods to a class like `DnsName`?

Reply:
Good point. Thanks. I was actually considering renaming the type from `DomainName` to be just `Name` (which is already under `Dns` namespace, so it's a `Dns::Name`). I think your point here makes a good argument why we should not call it `DomainName`. I will change it in the next push.

Reply:
Changed in new pushed commit.

Reply:
Thanks!